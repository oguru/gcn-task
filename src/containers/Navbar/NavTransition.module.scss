@import "../../variables.scss";

.enter {
    opacity: 0;
    overflow: hidden;
    transform: translateY(-100%);
}

.enterActive {
    opacity: 1;
    overflow: hidden;
    transform: translateY(0);
    transition-duration: $transition-speed;
}

.exit {
    opacity: 1;
    overflow: hidden;
    transform: translateY(0);
    transition: $transition-speed;
}

.exitActive {
    opacity: 0;
    overflow: hidden;
    transform: translateY(-150%);
}