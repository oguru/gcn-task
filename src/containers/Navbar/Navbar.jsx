import React, { useEffect, useState } from "react";
import { Link } from "@reach/router";
import styles from "./Navbar.module.scss";
import navTransition from "./NavTransition.module.scss";
import { CSSTransition } from "react-transition-group";
import NavItem from "../../components/NavItem";
import PropTypes from 'prop-types';
import gcnNavTop from "../../assets/GCNNavTop.png"
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faAngleRight } from '@fortawesome/free-solid-svg-icons'
import { faAngleLeft } from '@fortawesome/free-solid-svg-icons'

const Navbar = (props) => {

  // const [navLinkVisibility, toggleNavOpen] = useState(false);
  const [activeMenu, setActiveMenu] = useState("main");
  const [menuHeight, setMenuHeight] = useState(0);
  const [navOpen, setNavOpen] = useState(false);
  const [navJsx, setNavJsx] = useState({"main": 0});
  // const [navSublinks, setNavSublinks] = useState({"main": 1})

  useEffect(() => {
    const mainLinks = navMainLinks.map((link, index) => getLinks(link, index));
    setNavJsx({main: mainLinks});
    navMainLinks.forEach(link => getSubLinks(link));
 }, []);

  useEffect(() => {
    calcHeight();
  }, [activeMenu, navOpen]);

  const calcHeight = () => {
    const navItems = Object.keys(navJsx[activeMenu]);
    console.log(navItems);
    if (navItems) {
      const totalHeight = navItems.length * 30;
      setMenuHeight(totalHeight)
    }
  }

  const navIconAnim = navOpen ? "navCross" : "";

  const navMainLinks = [
    { name: "GCN Channels", childLinks: [
      { name: "GCN Show", path: "../category/gcn-show" },
      { name: "How To", path: "../category/how-to" },
      { name: "Maintenance", path: "../category/maintenance-mondays" },
      { name: "Ask GCN", path: "../category/ask-gcn" },
      { name: "Training", path: "../category/training" },
      { name: "Features", path: "../category/features" },
      { name: "Top 10s", path: "../category/top-10s" },
      { name: "GCN Racing", path: "../category/gcn-racing" },
      { name: "GCN Tech", path: "../category/gcn-tech" }
    ] },
    { name: "Presenters", childLinks: [
      {name: "About", path: "../presenters"},
      { name: "Conor Dunne", path: "../presenters/conor-dunne" },
      { name: "Daniel Lloyd", path: "../presenters/daniel-lloyd" },
      { name: "James Lowsley-Williams", path: "../presenters/james-lw" },
      { name: "Jon Cannings", path: "../presenters/jon-cannings" },
      { name: "Manon Lloyd", path: "../presenters/manon-lloyd" },
      { name: "Oliver Bridgewood", path: "../presenters/oliver-bridgewood" },
      { name: "Simon Richardson", path: "../presenters/simon-richardson" },
      { name: "Tom Last", path: "../presenters/tom-last" }      
    ] },
    { name: "Archives", path: "../archives" }
  ];

  const changeNav = (menuName) => {
    setActiveMenu(menuName.split(" ").join(""));
    // setMenuHeight("");
    // calcHeight();
  }

  const getSubLinks = (link) => {
    if (link.hasOwnProperty('childLinks')) {
      const subLinks = link.childLinks.map(subLink => getLinks(subLink));
      setNavJsx(navData => ({...navData, [link.name.split(" ").join("")]: subLinks}));
    };
  }

  const getLinks = (link, index) => {
    if (link.hasOwnProperty('childLinks')) {
      return (
        <li onClick={() => changeNav(link.name)} className={styles.navItemCont}>
          <p className={styles.navItem}>{link.name}</p>
          <span className={styles.navArrow}>
            <FontAwesomeIcon icon={faAngleRight}/>
          </span>
        </li>
      );
    }
    return (
      <li className={styles.navItemCont} onClick={() => setNavOpen(false)}>
        <Link
          className={styles.navItem}          
          key={link.name}
          // getProps={isActive}
          to={link.path}
        >
          {link.name}
        </Link>
        {/* <NavItem name={link.name} index={index} path={link.path} /> */}
      </li>
    );
  };

  const toggleNavOpen = () => {
    if (!navOpen) {
      setActiveMenu("main");
    }
    setNavOpen(!navOpen);
  }

  const backArrow = activeMenu === "main" ? "" : (
    <div className={styles.arrowCont} onClick={() => setActiveMenu("main")}>
      <span className={styles.navArrow} >
        <FontAwesomeIcon icon={faAngleLeft}/>
      </span>
    </div>
  );

  return (
    <>
      <nav className={styles.navbar}>
        <Link to={"/"}>
          <img src={gcnNavTop} alt="GCN Nav Top"></img>
        </Link>
        <div className={styles.navSecondary}>
          <div onClick={() => toggleNavOpen()} className={`${styles.burgerIcon}`}>
            <span className={styles[navIconAnim]}></span>
            <span className={styles[navIconAnim]}></span>
            <span className={styles[navIconAnim]}></span>
          </div>
        </div>
         <CSSTransition classNames={navTransition} in={navOpen} timeout={500} unmountOnExit >
          <div className={styles.navLinkBar} style={{height: menuHeight}}>
              {backArrow}
            <ul>
              {navJsx[activeMenu]}
            </ul>
          </div>
        </CSSTransition>
      </nav>
    </>
  );
};

export default Navbar;