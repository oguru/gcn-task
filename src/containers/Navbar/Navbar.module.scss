@import "../../variables.scss";

.navbar {
    width: 100%;
    
    img {
        display: block;
        position: relative;
        width: 100%;
        z-index: 1000;
    }

    .navLinkBar {
        background-color: black;
        display: flex;
        position: absolute;
        transition: $transition-speed ease-in-out;
        width: 100%;

        ul {
            flex-grow: 1;
            list-style: none;
            padding: 0;
        }

        .arrowCont {
            cursor: pointer;
            height: 100%;
            padding-right: $base-spacing;

            &:hover {
                background-color: $primary-color;
            }
        }

        .navArrow {
            align-items: center;
            color: white;
            display: flex;
            height: 30px;
            justify-content: center;
            margin-left: $base-spacing;

            svg {
               height: 20px;
               width: 20px; 
            }
        }
    }

    .navLinksOpen {
        padding: 0 $base-spacing $base-spacing;        
    }

    .navSecondary {
        align-items: center;
        background-color: black;
        display: flex;
        height: 36px;
        justify-content: space-between;
        padding: $base-spacing/2 $base-spacing;
        position: relative;
        z-index: 1000;

        .burgerIcon {
            cursor: pointer;
            display: flex;
            flex-direction: column;
            justify-content: center;
            height: 15px;
            position: relative;
            width: 20px;
            
            span {
                background-color: white;
                border-radius: $base-spacing;
                display: block;
                height: 2px;
                position: absolute;
                top: 0;
                transform: rotate(0deg) translate(0px);
                transition: $transition-speed ease-in-out;
                width: 20px;
            }

            span:nth-child(2) {
                top: 7.5px;
            }

            span:nth-child(3) {
                top: 15px;
            }

            span.navCross:nth-child(1) {
                height: 3px;
                top: 7.5px;
                transform: rotate(135deg);
            }

            span.navCross:nth-child(2) {
                opacity: 0;
                transform: translateX(-100%);
            }

            span.navCross:nth-child(3) {
                height: 3px;
                top: 7.5px;
                transform: rotate(-135deg);
            }
        }

        .socialMediaIcons {
            display: flex;
        }
    }
}
