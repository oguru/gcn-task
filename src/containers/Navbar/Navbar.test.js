import { mount } from "enzyme";
import { render, cleanup } from "@testing-library/react";
import React from "react";
import Navbar from "./Navbar";

describe("Navbar tests", () => {
  let component;
  
  beforeEach(() => {
    component = mount(<Navbar />);
  });
  
  afterEach(cleanup);
  
  it("should render", () => {
    expect(render(<Navbar />)).toBeTruthy();
  });

  it("should match the snapshot", () => {
    expect(component).toMatchSnapshot();
  });

  it("should contain a nav tag", () => {
    expect(component.find("nav").length).toEqual(1);
  });

  it("should have a burger menu icon", () => {
    expect(component.find(".burgerIcon").length).toBe(1);
  })

  it("should add navbar links when burger menu is clicked", () => {
    expect(component.find(".navLinkBar").length).toBe(0);
    component.find(".burgerIcon").simulate("click");
    expect(component.find(".navLinkBar").length).toBe(1);
  })
});
