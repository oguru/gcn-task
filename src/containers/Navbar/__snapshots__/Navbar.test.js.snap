// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar tests should match the snapshot 1`] = `ReactWrapper {}`;
